### PR TITLE
New properties API

### PR DESCRIPTION
**Pull request by robertbastian**
#3573

### CODE REVIEW

**Comment by robertbastian** [comment]
Will fix docs tests after initial review.

File: components/properties/src/props.rs
Comment (Manishearth) [inline components/properties/src/props.rs:69]: issue: I'd rather not do this; these types are tagged as unstable in the docs.

Can we instead just use regular integers in the data and convert here? we erase the type anyway

Comment (robertbastian) [reply components/properties/src/props.rs:69]: I think we should consider them stable instead.

Comment (robertbastian) [reply components/properties/src/props.rs:69]: The reason I moved them is that they have `make_ule`, and the ULE types should not be stable.

**Comment by robertbastian** [comment]
https://icu4x-pr-artifacts.storage.googleapis.com/gha/3d38f6a4faa482c01d0d8f146139a628e5ffe9d9/rustdoc/icu/properties/index.html

File: components/properties/src/props.rs
Comment (Manishearth) [reply components/properties/src/props.rs:78]: @robertbastian Right, I wish to also consider them stable, which is why I don't want them in the provider module.


Don't worry about `make_ule`, those are going to go away before 1.0 https://github.com/unicode-org/icu4x/issues/5127 . I've been waiting for this PR to land to make that change.

**Review by Manishearth — COMMENTED** [review]
Overall happy with the direction here, mostly have comments on code organization:

 - I think the properties should all canonically live in `crate::props`, not under `provider`. Don't worry about the ULE types, they're going away
 - I think enumerated properties should automatically generate names

Since this is a large PR and unwieldy to review I'm okay merging this with followups.

**Review by Manishearth — APPROVED** [review]
Approving with the two code organization bits as followups. I'm happy to do the props one since I want to touch that code anyway.

**Review by Manishearth — APPROVED** [review]

**Comment by robertbastian** [comment]
> I think enumerated properties should automatically generate names

If you mean merging `EnumeratedProperty`, `NamedEnumeratedProperty` and `ParseableEnumeratedProperty`, that doesn't work because 
* `Script` doesn't use `NamedEnumeratedProperty`/`PropertyNames`, it uses `ScriptMapper` to return `icu::locale::subtags::Script`
* `GeneralCategoryGroup` is a pseudo-property, it doesn't implement `EnumeratedProperty`, but does implement `ParseableEnumeratedProperty`

**Comment by Manishearth** [comment]
> If you mean merging `EnumeratedProperty`, `NamedEnumeratedProperty` and `ParseableEnumeratedProperty`, that doesn't work because

I meant having the macro generate names too. Not merging the traits.

**Comment by Manishearth** [comment]
> I think the properties should all canonically live in crate::props, not under provider. Don't worry about the ULE types, they're going away


I'll take on this followup

**Comment by Manishearth** [comment]
Done in #5555. GeneralCategoryULE's macro sticks around since we've not done #5127 yet, but in the longer run we can get rid of it with a macro.

**Pull request merged**
